scalajslib/worker/1/src/ScalaJSWorkerImpl.scala

```diff
@@ -94,12 +96,49 @@ class ScalaJSWorkerImpl extends ScalaJSWorkerApi {
         else withESVersion_1_5_minus(scalaJSESFeatures)
 
       val useClosure = input.fullOpt && input.moduleKind != ModuleKind.ESModule
-      val config = StandardConfig()
+      var config = StandardConfig()
         .withOptimizer(input.fullOpt)
         .withClosureCompilerIfAvailable(useClosure)
         .withSemantics(semantics)
         .withModuleKind(scalaJSModuleKind)
         .withESFeatures(scalaJSESFeatures)
+
+      // Separating ModuleSplitStyle.SmallModulesFor in a standalone object
+      // avoids early classloading which fails in Scala.js 1.9- where this
+      // class doesn't exist
+      object ScalaJSSmallModulesFor {
+        def apply(packages: List[String]): ScalaJSModuleSplitStyle =
+          ScalaJSModuleSplitStyle.SmallModulesFor(packages)
+      }
+
+      def withModuleSplitStyle_1_3_plus(config: StandardConfig): StandardConfig = {
+        config.withModuleSplitStyle(
+          input.moduleSplitStyle match {
+            case ModuleSplitStyle.FewestModules => ScalaJSModuleSplitStyle.FewestModules
+            case ModuleSplitStyle.SmallestModules => ScalaJSModuleSplitStyle.SmallestModules
+            case v @ ModuleSplitStyle.SmallModulesFor(packages) =>
+              if (minorIsGreaterThanOrEqual(10)) ScalaJSSmallModulesFor(packages)
+              else throw new Exception(
+                s"ModuleSplitStyle $v is not supported with Scala.js < 1.10. Either update Scala.js or use one of ModuleSplitStyle.SmallestModules or ModuleSplitStyle.FewestModules"
+              )
+          }
+        )
+      }
+
+      def withModuleSplitStyle_1_2_minus(config: StandardConfig): StandardConfig = {
+        input.moduleSplitStyle match {
+          case ModuleSplitStyle.FewestModules =>
+          case v => throw new Exception(
+              s"ModuleSplitStyle $v is not supported with Scala.js < 1.2. Either update Scala.js or use ModuleSplitStyle.FewestModules"
+            )
+        }
+        config
+      }
+
+      config =
+        if (minorIsGreaterThanOrEqual(3)) withModuleSplitStyle_1_3_plus(config)
+        else withModuleSplitStyle_1_2_minus(config)
+
       StandardImpl.linker(config)
```

Comment:
Could we use a new `val` here, to avoid the `var` above?

Reply:
Done ✅